cdap-app-fabric/src/main/java/co/cask/cdap/scheduler/NotificationSubscriberService.java

```diff
@@ -0,0 +1,359 @@
+  // Sampling log only log once per 10000
+  private static final Logger SAMPLING_LOG = Loggers.sampling(LOG, LogSamplers.limitRate(10000));
+  private static final Gson GSON = new Gson();
+
+  private final List<NotificationSubscriberThread> subscriberThreads;
+  /** Temporarily unused **/
+  private final Transactional transactional;
+  private final MultiThreadMessagingContext messagingContext;
+  private final DatasetFramework datasetFramework;
+  private final Store store;
+  private final ProgramLifecycleService lifecycleService;
+  private final PropertiesResolver propertiesResolver;
+  private final NamespaceQueryAdmin namespaceQueryAdmin;
+  private final CConfiguration cConf;
+  private ScheduleTaskRunner taskRunner;
+  private ListeningExecutorService taskExecutorService;
+
+  @Inject
+  NotificationSubscriberService(MessagingService messagingService,
+                                Store store,
+                                ProgramLifecycleService lifecycleService, PropertiesResolver propertiesResolver,
+                                NamespaceQueryAdmin namespaceQueryAdmin,
+                                CConfiguration cConf,
+                                DatasetFramework datasetFramework,
+                                TransactionSystemClient txClient) {
+    this.store = store;
+    this.lifecycleService = lifecycleService;
+    this.propertiesResolver = propertiesResolver;
+    this.namespaceQueryAdmin = namespaceQueryAdmin;
+    this.cConf = cConf;
+    this.messagingContext = new MultiThreadMessagingContext(messagingService);
+    this.transactional = Transactions.createTransactionalWithRetry(
+      Transactions.createTransactional(new MultiThreadDatasetCache(
+        new SystemDatasetInstantiator(datasetFramework), txClient,
+        NamespaceId.SYSTEM, ImmutableMap.<String, String>of(), null, null, this.messagingContext)),
+      RetryStrategies.retryOnConflict(20, 100)
+    );
+    this.datasetFramework = datasetFramework;
+    this.subscriberThreads = new ArrayList<>();
+  }
+
+  @Override
+  protected void startUp() throws Exception {
+    taskExecutorService = MoreExecutors.listeningDecorator(
+      Executors.newCachedThreadPool(Threads.createDaemonThreadFactory("notification-subscriber-task")));
+    taskRunner = new ScheduleTaskRunner(store, lifecycleService, propertiesResolver,
+                                        taskExecutorService, namespaceQueryAdmin, cConf);
+  }
+
+  @Override
+  protected void run() {
+    LOG.info("Start running NotificationSubscriberService");
+    if (!isRunning()) {
+      return;
+    }
+
+    subscriberThreads.add(new TimeEventNotificationSubscriberThread(null));
+    subscriberThreads.add(new DataEventNotificationSubscriberThread(null));
+
+    for (NotificationSubscriberThread thread : subscriberThreads) {
+      thread.start();
+    }
+
+    for (NotificationSubscriberThread thread : subscriberThreads) {
+      try {
+        thread.join();
```

Comment:
You can use `Uninterruptibles.joinUninterruptibly` to avoid the unnecessary try-catch since the run thread is not getting interrupted.